tests/common/test_host_basics.hpp

```diff
@@ -534,6 +534,227 @@ TYPED_TEST_P(test_host_basics, local_transformations) {
+  ASSERT_NEAR(local[0], 2., this->m_isclose);
+  ASSERT_NEAR(local[1], 4., this->m_isclose);
+  ASSERT_NEAR(local[2], 1., this->m_isclose);
+
+  // Local to global transformation
+  const typename TypeParam::point3 global2 = c3.local_to_global(trf, local);
+
+  // Check if the same global position is obtained
+  ASSERT_NEAR(global1[0], global2[0], this->m_isclose);
+  ASSERT_NEAR(global1[1], global2[1], this->m_isclose);
+  ASSERT_NEAR(global1[2], global2[2], this->m_isclose);
+}
+
+// This test polar2 coordinate
+TYPED_TEST_P(test_host_basics, polar2) {
+
+  // Preparation work
+  const typename TypeParam::vector3 z =
+      typename TypeParam::vector_actor().normalize(
+          typename TypeParam::vector3{0., 0., 1.});
+  const typename TypeParam::vector3 x =
+      typename TypeParam::vector_actor().normalize(
+          typename TypeParam::vector3{1., 0., 0.});
+  const typename TypeParam::point3 t = {2., 3., 4.};
+  const typename TypeParam::transform3 trf(t, z, x);
+  const typename TypeParam::polar2 p2;
+
+  // Global position on surface
+  const typename TypeParam::point3 global1 = {4., 7., 4.};
+
+  // Global to local transformation
+  const typename TypeParam::point2 local = p2.global_to_local(trf, global1);
+
+  // Check if the local position is correct
+  ASSERT_NEAR(local[0], std::sqrt(20.), this->m_isclose);
+  ASSERT_NEAR(local[1], atan2(4., 2.), this->m_isclose);
+
+  // Local to global transformation
+  const typename TypeParam::point3 global2 = p2.local_to_global(trf, local);
+
+  // Check if the same global position is obtained
+  ASSERT_NEAR(global1[0], global2[0], this->m_isclose);
+  ASSERT_NEAR(global1[1], global2[1], this->m_isclose);
+  ASSERT_NEAR(global1[2], global2[2], this->m_isclose);
+}
+
+// This test cylindrical2 coordinate
+TYPED_TEST_P(test_host_basics, cylindrical2) {
+
+  // Preparation work
+  const typename TypeParam::vector3 z =
+      typename TypeParam::vector_actor().normalize(
+          typename TypeParam::vector3{0., 0., 1.});
+  const typename TypeParam::vector3 x =
+      typename TypeParam::vector_actor().normalize(
+          typename TypeParam::vector3{1., 0., 0.});
+  const typename TypeParam::point3 t = {2., 3., 4.};
+  const typename TypeParam::transform3 trf(t, z, x);
+  const typename TypeParam::cylindrical2 c2;
+
+  // Define cylinder mask
+  struct cylinder_mask {
+    typename TypeParam::scalar r = 0.;
+    typename TypeParam::scalar radius() const { return r; }
+  };
+
+  const typename TypeParam::scalar r = 2.;
+  const cylinder_mask mask{r};
+
+  // Global position on surface
+  const typename TypeParam::point3 global1 = {
+      typename TypeParam::scalar{3.4142136},
+      typename TypeParam::scalar{4.4142136}, typename TypeParam::scalar{9.}};
+
+  // Global to local transformation
+  const typename TypeParam::point2 local = c2.global_to_local(trf, global1);
+
+  // Check if the local position is correct
+  ASSERT_NEAR(local[0], r * M_PI / 4., this->m_isclose);
```

Comment:
And then here you are using ```doubles``` all the way. I am not sure how much consistency we want to require for this. Is this actually important or am I mostly pedantic in the testing code?

Reply:
I think it is OK as I don't see any narrowing conversion warnings

Reply:
As far as I know, you will not receive many narrowing warnings on implicit conversions without using curly braces. In this case, however, I guess it will not give you a warning anyway. I don't think that comparisons will trigger a warning (or whatever google test is doing)?

Reply:
OK. I put the scalar conversion to `M_PI_4` now. Please tell me explicitly if it needs to be fixed.

Reply:
I am not quite following, but if you want to be on the safe side, I guess you could static cast the ```double``` to a ```scalar_type```. That this still implicitly casts ```r``` to ```double``` doesn't matter so much in tests I would say
```
static_cast<scalar_type>(r * M_PI / 4.)
```

Reply:
The test should work fine anyway, so, I'll leave it up to you